message_ix_models/tools/costs/decay.py

```diff
@@ -7,180 +9,308 @@
 from .regional_differentiation import get_raw_technology_mapping, subset_module_map
 
 
-def get_cost_reduction_data(module) -> pd.DataFrame:
-    """Get cost reduction data from file.
-
-    Raw data on cost reduction in 2100 for technologies are read from
-    :file:`data/[module]/cost_reduction_[module].csv`, based on GEA data.
+def _get_module_scenarios_reduction(module, energy_map_df, tech_map_df):
+    """Get scenarios reduction categories for a module.
 
     Parameters
     ----------
     module : str
-        Model module
+        The module for which to get scenarios reduction categories.
+    energy_map_df : pandas.DataFrame
+        The technology mapping for the energy module.
+    tech_map_df : pandas.DataFrame
+        The technology mapping for the specific module.
 
     Returns
     -------
     pandas.DataFrame
         DataFrame with columns:
-
         - message_technology: name of technology in MESSAGEix
-        - reduction_rate: the cost reduction rate (either very_low, low, medium, high,
-        or very_high)
-        - cost_reduction: cost reduction in 2100 (%)
+        - SSP1: scenario reduction category for SSP1
+        - SSP2: scenario reduction category for SSP2
+        - SSP3: scenario reduction category for SSP3
+        - SSP4: scenario reduction category for SSP4
+        - SSP5: scenario reduction category for SSP5
+        - LED: scenario reduction category for LED
     """
+    # Get reduction scenarios for energy module
+    scenarios_energy = pd.read_csv(
+        package_data_path("costs", "energy", "scenarios_reduction.csv")
+    )
 
-    # Get full list of technologies from mapping
-    tech_map = energy_map = get_raw_technology_mapping("energy")
+    # for technologies in energy_map that are not in scenarios_energy,
+    # assume scenario reduction across all scenarios is "none"
+    # add same columns as scenarios_energy
+    # and set all values to "none" except for message_technology column
+    scenarios_energy_no_reduction = energy_map_df.query(
+        "message_technology not in @scenarios_energy.message_technology"
+    )[["message_technology"]].assign(
+        **{
+            col: "none"
```

Comment:
The "none" is meant to be a category (as opposed to "low", "high", "medium", etc), and later on all instances of a "none" reduction category is assigned to a value of 0.

```
df["cost_reduction"] = np.where(df.reduction_rate == "none", 0, df.cost_reduction)
```

I'm not familiar enough with `None` objects to know if it could replace a "none" string in this case?

Reply:
Ah, I see. If this is its own category, probably best to keep it a string. Using `None` instead would be indistinguishable from the variable having no (the `None`) value.